test/blocks/featured-article/featured-article.test.js

```diff
@@ -0,0 +1,25 @@
+/* eslint-disable no-unused-expressions */
+/* global describe it */
+
+import { readFile } from '@web/test-runner-commands';
+import { expect } from '@esm-bundle/chai';
+
+document.body.innerHTML = await readFile({ path: './mocks/body.html' });
+const { default: init } = await import('../../../libs/blocks/featured-article/featured-article.js')
+
+describe('creates feature article block', async () => {
+  await init(document);
```

Comment:
yeah! - here we go.

Tbh in my opinion you should aim for full test coverage of the featured-article.js file. You can achieve this pretty easily by passing getBlogArticle into the decorateFeaturedArticle method as a parameter, and then mocking getBlogArticle . 
I'm not sure what the others in the team would think about us slightly adapting code to make it easier to test though. 
@chrischrischris, @auniverseaway , @rgclayton what do you guys think?

Reply:
I just realized there's a much better way to test this @saugatmalla 😅 
You could just mock the call like this in your test file:
```
/* eslint-disable no-unused-expressions */
/* global describe it */

import { readFile } from '@web/test-runner-commands';
import { expect } from '@esm-bundle/chai';
import { stub } from 'sinon';

const articleHtml = await readFile({ path: './mocks/article.html' });
const ogFetch = window.fetch;
window.fetch = stub();
window.fetch.withArgs(`https://main--blog--adobecom.hlx.page/published/2022/11/15/voices-of-our-community`).returns(
  new Promise((resolve) => {
    resolve({
      ok: true,
      text: function() {
        return articleHtml
      }
    });
  }),
);
const restoreFetch = () => {
  window.fetch = ogFetch;
};

document.body.innerHTML = await readFile({ path: './mocks/body.html' });
const { default: init } = await import('../../../libs/blocks/featured-article/featured-article.js')

describe('creates feature article block', async () => {
  before(async () => {
    await init(document);
  });
  after(() => {
    restoreFetch()
  });

  it('has card with url', () => {
    const card = document.querySelector('.featured-article-card');
    expect(card).to.exist;
  });
  it('has featured article image', () => {
    const image = document.querySelector('.featured-article-card-image');
    expect(image).to.exist;
  });
  it('has content', () => {
    const content = document.querySelector('.featured-article-card-body');
    expect(content).to.exist;
  });
});

```

and then add another html file in your mocks folder called article.html with this content: 
```
<!DOCTYPE html>
<html>
<head>
  <title>"In a New York Minute” with Maliyamungu Muhande and Imani Dennison | Adobe Blog</title>
  <link rel="canonical" href="https://main--blog--adobecom.hlx.page/published/2022/11/15/voices-of-our-community">
  <meta name="description" content="Filmmakers Maliyamungu Muhande and Imani Dennison collaborated together with Muhande creating an original short film around the theme of “community”, and Dennison documenting and filming Muhande’s creative process behind the scenes.">
  <meta property="og:title" content="&#x22;In a New York Minute” with Maliyamungu Muhande and Imani Dennison  | Adobe Blog">
  <meta property="og:description" content="Filmmakers Maliyamungu Muhande and Imani Dennison collaborated together with Muhande creating an original short film around the theme of “community”, and Dennison documenting and filming Muhande’s creative process behind the scenes.">
  <meta property="og:url" content="https://main--blog--adobecom.hlx.page/published/2022/11/15/voices-of-our-community">
  <meta property="og:image" content="https://main--blog--adobecom.hlx.page/published/2022/11/15/media_1cd7a8c416aaff1e70d71679b74441e46c4c545f3.png?width=1200&#x26;format=pjpg&#x26;optimize=medium">
  <meta property="og:image:secure_url" content="https://main--blog--adobecom.hlx.page/published/2022/11/15/media_1cd7a8c416aaff1e70d71679b74441e46c4c545f3.png?width=1200&#x26;format=pjpg&#x26;optimize=medium">
  <meta property="og:image:alt" content="Still from &#x27;New York Minute&#x27;">
  <meta property="article:tag" content="Creativity">
  <meta property="article:tag" content="Creative Cloud">
  <meta property="article:tag" content="Corporate Alignment">
  <meta property="article:tag" content="Brand">
  <meta property="article:tag" content="Diverse Voices">
  <meta property="article:tag" content="Brand Strategy">
</head>
<body>
<header></header>
<main>
</main>
<footer></footer>
</body>
</html>
```

Reply:
> window.fetch.withArgs(`https://main--blog--adobecom.hlx.page/published/2022/11/15/voices-of-our-community`).returns(
>   new Promise((resolve) => {
>     resolve({
>       ok: true,
>       text: function() {
>         return articleHtml
>       }
>     });
>   }),
> );

Hey @vhargrave! Thank you for helping with this test file, I seriously needed it. I copy pasted with a minor fix.
With this approach, we are using stub on fetch to return our mock data, can we stub the `getBlogArticle` & `buildArticleCard` methods of the `decorateFeaturedArticle`? If yes, how could we achieve that?

Reply:
My pleasure! The stubs are a little tricky, especially with HTTP Requests - so glad that it helped.

No, you can't stub those methods unfortunately, because they're being imported. 
If you did want to stub them, you'd have to pass those methods as parameters to the decorateFeaturedArticle method. 
then you'd create your stubs kind of like this: 
```
const getBlogArticle = stub().returns('whatever you want');
const buildArticleCard = stub().returns('dito');
decorateFeaturedArticle(getBlogArticle, buildArticleCard, ...all the other params);
```

This was the question that I asked the rest of the team at the beginning of this thread. It feels wrong changing a method just to make it easier to test, so I'd leave it be for now. With your unit tests and fetch stub you're getting full test coverage now either way.